vertx-graphql fails in native mode - GraphQLInputDeserializer has no default (no arg) constructor

integration-tests/vertx-graphql fails in native mode - GraphQLInputDeserializer has no default (no arg) constructor

```
Caused by: java.lang.IllegalArgumentException: Class
  io.vertx.ext.web.handler.graphql.impl.GraphQLInputDeserializer
  has no default (no arg) constructor
	at com.fasterxml.jackson.databind.util.ClassUtil.createInstance(ClassUtil.java:554)
	
```

CC @machi1990 @cristianonicolai @gsmet 

```
2019-11-06 12:35:29,018 INFO  [io.quarkus] (main) Installed features: [cdi, vertx-graphql]
2019-11-06 12:35:29,876 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-3) HTTP Request to /graphql failed, error id: bc94e483-9cae-4472-aa5c-bf8d07dbc6d9-1: io.vertx.core.json.DecodeException: Failed to decode:Class io.vertx.ext.web.handler.graphql.impl.GraphQLInputDeserializer has no default (no arg) constructor
 at [Source: (io.netty.buffer.ByteBufInputStream); line: 1, column: 1]
	at io.vertx.core.json.jackson.JacksonCodec.fromParser(JacksonCodec.java:97)
	at io.vertx.core.json.jackson.JacksonCodec.fromBuffer(JacksonCodec.java:66)
	at io.vertx.ext.web.handler.graphql.impl.GraphQLHandlerImpl.handlePostJson(GraphQLHandlerImpl.java:141)
	at io.vertx.ext.web.handler.graphql.impl.GraphQLHandlerImpl.handlePost(GraphQLHandlerImpl.java:128)
	at io.vertx.ext.web.handler.graphql.impl.GraphQLHandlerImpl.lambda$handle$2(GraphQLHandlerImpl.java:86)
	at io.vertx.core.http.HttpServerRequest.lambda$bodyHandler$0(HttpServerRequest.java:216)
	at io.vertx.core.http.impl.HttpServerRequestImpl.onEnd(HttpServerRequestImpl.java:530)
	at io.vertx.core.http.impl.HttpServerRequestImpl.lambda$pendingQueue$1(HttpServerRequestImpl.java:118)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:237)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:127)
	at io.vertx.core.http.impl.HttpServerRequestImpl.handleEnd(HttpServerRequestImpl.java:514)
	at io.vertx.core.http.impl.Http1xServerConnection.handleEnd(Http1xServerConnection.java:176)
	at io.vertx.core.http.impl.Http1xServerConnection.handleContent(Http1xServerConnection.java:163)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:140)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:369)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:232)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:173)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:102)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:287)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:328)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:302)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.vertx.core.http.impl.Http1xOrH2CHandler.end(Http1xOrH2CHandler.java:61)
	at io.vertx.core.http.impl.Http1xOrH2CHandler.channelRead(Http1xOrH2CHandler.java:38)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:287)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1422)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:931)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:700)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:635)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:552)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:514)
	at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1044)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:460)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Class io.vertx.ext.web.handler.graphql.impl.GraphQLInputDeserializer has no default (no arg) constructor
 at [Source: (io.netty.buffer.ByteBufInputStream); line: 1, column: 1]
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:268)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
	at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:477)
	at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer(ObjectMapper.java:4191)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3980)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2277)
	at io.vertx.core.json.jackson.JacksonCodec.fromParser(JacksonCodec.java:95)
	... 58 more
Caused by: java.lang.IllegalArgumentException: Class io.vertx.ext.web.handler.graphql.impl.GraphQLInputDeserializer has no default (no arg) constructor
	at com.fasterxml.jackson.databind.util.ClassUtil.createInstance(ClassUtil.java:554)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.deserializerInstance(DefaultDeserializationContext.java:229)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findDeserializerFromAnnotation(DeserializerCache.java:427)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:326)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264)
	... 65 more

```

--- a/ci-templates/stages.yml
+++ b/ci-templates/stages.yml
@@ -310,11 +310,12 @@ stages:
         parameters:
           poolSettings: ${{parameters.poolSettings}}
           expectUseVMs: ${{parameters.expectUseVMs}}
-          timeoutInMinutes: 25
+          timeoutInMinutes: 30
           modules:
             - resteasy-jackson
             - vertx
             - vertx-http
+            - vertx-graphql
             - virtual-http
           name: http
 

--- a/core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/NativeImageResourceDirectoryBuildItem.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/NativeImageResourceDirectoryBuildItem.java
@@ -1,0 +1,19 @@
+package io.quarkus.deployment.builditem.nativeimage;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * A build item that indicates that directory resources should be included in the native image
+ */
+public final class NativeImageResourceDirectoryBuildItem extends MultiBuildItem {
+
+    private final String path;
+
+    public NativeImageResourceDirectoryBuildItem(String path) {
+        this.path = path;
+    }
+
+    public String getPath() {
+        return path;
+    }
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java
@@ -2,17 +2,26 @@ package io.quarkus.deployment.steps;
 
 import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
 
+import java.io.File;
+import java.io.IOException;
 import java.lang.reflect.AccessibleObject;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Executable;
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Enumeration;
 import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
 import java.util.stream.Collectors;
 
 import org.graalvm.nativeimage.ImageSingletons;
@@ -22,10 +31,12 @@ import org.graalvm.nativeimage.impl.RuntimeClassInitializationSupport;
 
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.DeploymentClassLoaderBuildItem;
 import io.quarkus.deployment.builditem.GeneratedNativeImageClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.NativeImageProxyDefinitionBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBundleBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceDirectoryBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveFieldBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveMethodBuildItem;
@@ -55,6 +66,32 @@ public class NativeImageAutoFeatureStep {
     static final String BEFORE_ANALYSIS_ACCESS = Feature.BeforeAnalysisAccess.class.getName();
     static final String DYNAMIC_PROXY_REGISTRY = "com.oracle.svm.core.jdk.proxy.DynamicProxyRegistry";
     static final String LOCALIZATION_SUPPORT = "com.oracle.svm.core.jdk.LocalizationSupport";
+
+    @BuildStep
+    List<NativeImageResourceBuildItem> registerPackageResources(
+            List<NativeImageResourceDirectoryBuildItem> nativeImageResourceDirectories,
+            DeploymentClassLoaderBuildItem classLoader)
+            throws IOException, URISyntaxException {
+        List<NativeImageResourceBuildItem> resources = new ArrayList<>();
+
+        for (NativeImageResourceDirectoryBuildItem nativeImageResourceDirectory : nativeImageResourceDirectories) {
+            String path = classLoader.getClassLoader().getResource(nativeImageResourceDirectory.getPath()).getPath();
+            File resourceFile = Paths.get(new URL(path.substring(0, path.indexOf("!"))).toURI()).toFile();
+            try (JarFile jarFile = new JarFile(resourceFile)) {
+                Enumeration<JarEntry> entries = jarFile.entries();
+                while (entries.hasMoreElements()) {
+                    JarEntry entry = entries.nextElement();
+                    String resourceName = entry.getName();
+                    if (!entry.isDirectory() && resourceName.startsWith(nativeImageResourceDirectory.getPath())
+                            && !resourceName.endsWith(".class")) {
+                        resources.add(new NativeImageResourceBuildItem(resourceName));
+                    }
+                }
+            }
+        }
+
+        return resources;
+    }
 
     @BuildStep
     void generateFeature(BuildProducer<GeneratedNativeImageClassBuildItem> nativeImageClass,

--- a/docs/src/main/asciidoc/writing-extensions.adoc
+++ b/docs/src/main/asciidoc/writing-extensions.adoc
@@ -1642,6 +1642,9 @@ executable. Some of these build items are listed below:
 `io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBuildItem`::
  Includes static resources into the native executable.
 
+`io.quarkus.deployment.builditem.nativeimage.NativeImageResourceDirectoryBuildItem`::
+ Includes directory's static resources into the native executable.
+
 `io.quarkus.deployment.builditem.nativeimage.RuntimeReinitializedClassBuildItem`::
 A class that will be reinitialized at runtime by Substrate. This will result in the static initializer running twice.
 
@@ -2099,16 +2102,16 @@ public final class MyExtProcessor {
     @BuildStep
     void registerNativeImageReources(BuildProducer<ServiceProviderBuildItem> services) {
         String service = "META-INF/services/" + io.quarkus.SomeService.class.getName();
-        
+
         // find out all the implementation classes listed in the service files
-        Set<String> implementations = 
+        Set<String> implementations =
             ServiceUtil.classNamesNamedIn(Thread.currentThread().getContextClassLoader(),
                                           service);
 
-        // register every listed implementation class so they can be instantiated 
+        // register every listed implementation class so they can be instantiated
         // in native-image at run-time
         services.produce(
-            new ServiceProviderBuildItem(io.quarkus.SomeService.class.getName(), 
+            new ServiceProviderBuildItem(io.quarkus.SomeService.class.getName(),
                                          implementations.toArray(new String[0])));
     }
 }
@@ -2130,13 +2133,13 @@ public final class MyExtProcessor {
     void registerNativeImageReources(BuildProducer<NativeImageResourceBuildItem> resource,
                                      BuildProducer<ReflectiveClassBuildItem> reflectionClasses) {
         String service = "META-INF/services/" + io.quarkus.SomeService.class.getName();
-        
+
         // register the service file so it is visible in native-image
         resource.produce(new NativeImageResourceBuildItem(service));
-        
-        // register every listed implementation class so they can be inspected/instantiated 
+
+        // register every listed implementation class so they can be inspected/instantiated
         // in native-image at run-time
-        Set<String> implementations = 
+        Set<String> implementations =
             ServiceUtil.classNamesNamedIn(Thread.currentThread().getContextClassLoader(),
                                           service);
         reflectionClasses.produce(
@@ -2148,7 +2151,7 @@ public final class MyExtProcessor {
 While this is the easiest way to get your services running natively, it's less efficient than scanning the implementation
 classes at build time and generating code that registers them at static-init time instead of relying on reflection.
 
-You can achieve that by adapting the previous build step to use a static-init recorder instead of registering 
+You can achieve that by adapting the previous build step to use a static-init recorder instead of registering
 classes for reflection:
 
 [source,java]
@@ -2157,19 +2160,19 @@ public final class MyExtProcessor {
 
     @BuildStep
     @Record(ExecutionTime.STATIC_INIT)
-    void registerNativeImageReources(RecorderContext recorderContext, 
+    void registerNativeImageReources(RecorderContext recorderContext,
                                      SomeServiceRecorder recorder) {
         String service = "META-INF/services/" + io.quarkus.SomeService.class.getName();
-        
+
         // read the implementation classes
         Collection<Class<? extends io.quarkus.SomeService>> implementationClasses = new LinkedHashSet<>();
         Set<String> implementations = ServiceUtil.classNamesNamedIn(Thread.currentThread().getContextClassLoader(),
                                                                     service);
         for(String implementation : implementations) {
-            implementationClasses.add((Class<? extends io.quarkus.SomeService>) 
+            implementationClasses.add((Class<? extends io.quarkus.SomeService>)
                 recorderContext.classProxy(implementation));
         }
-        
+
         // produce a static-initializer with those classes
         recorder.configure(implementationClasses);
     }

--- a/extensions/mailer/runtime/src/test/java/io/quarkus/mailer/runtime/MailerImplTest.java
+++ b/extensions/mailer/runtime/src/test/java/io/quarkus/mailer/runtime/MailerImplTest.java
@@ -48,7 +48,7 @@ class MailerImplTest {
     @AfterAll
     static void stopWiser() {
         wiser.stop();
-        vertx.close();
+        vertx.close().toCompletableFuture().join();
     }
 
     @BeforeEach

--- a/extensions/mailer/runtime/src/test/java/io/quarkus/mailer/runtime/MockMailerImplTest.java
+++ b/extensions/mailer/runtime/src/test/java/io/quarkus/mailer/runtime/MockMailerImplTest.java
@@ -32,7 +32,7 @@ class MockMailerImplTest {
 
     @AfterAll
     static void stop() {
-        vertx.close();
+        vertx.close().toCompletableFuture().join();
     }
 
     @BeforeEach

--- a/extensions/vertx-graphql/deployment/pom.xml
+++ b/extensions/vertx-graphql/deployment/pom.xml
@@ -26,6 +26,16 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-vertx-graphql</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5-internal</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>rest-assured</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlConfig.java
+++ b/extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlConfig.java
@@ -1,0 +1,32 @@
+package io.quarkus.vertx.graphql.deployment;
+
+import io.quarkus.runtime.annotations.ConfigGroup;
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot
+public final class VertxGraphqlConfig {
+    /**
+     * GraphQL UI configuration
+     */
+    @ConfigItem
+    VertxGraphqlUiConfig ui;
+
+    @ConfigGroup
+    public static class VertxGraphqlUiConfig {
+        /**
+         * If GraphQL UI should be included every time. By default this is only included when the application is running
+         * in dev mode.
+         */
+        @ConfigItem(defaultValue = "false")
+        boolean alwaysInclude;
+
+        /**
+         * The path where GraphQL UI is available.
+         * <p>
+         * The value `/` is not allowed as it blocks the application from serving anything else.
+         */
+        @ConfigItem(defaultValue = "/graphql-ui")
+        String path;
+    }
+}

--- a/extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlProcessor.java
+++ b/extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlProcessor.java
@@ -2,17 +2,31 @@ package io.quarkus.vertx.graphql.deployment;
 
 import java.util.Arrays;
 import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
+import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.ExecutionTime;
+import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
-import io.quarkus.deployment.builditem.nativeimage.NativeImagePackageResourceBuildItem;
+import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceDirectoryBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.runtime.configuration.ConfigurationException;
+import io.quarkus.vertx.graphql.runtime.VertxGraphqlRecorder;
+import io.quarkus.vertx.http.deployment.RouteBuildItem;
 import io.quarkus.vertx.http.deployment.WebsocketSubProtocolsBuildItem;
+import io.quarkus.vertx.http.deployment.devmode.NotFoundPageDisplayableEndpointBuildItem;
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
 import io.vertx.ext.web.handler.graphql.impl.GraphQLBatch;
 import io.vertx.ext.web.handler.graphql.impl.GraphQLInputDeserializer;
 import io.vertx.ext.web.handler.graphql.impl.GraphQLQuery;
 
 class VertxGraphqlProcessor {
+    private static Pattern TRAILING_SLASH_SUFFIX_REGEX = Pattern.compile("/+$");
+
     @BuildStep
     FeatureBuildItem feature() {
         return new FeatureBuildItem(FeatureBuildItem.VERTX_GRAPHQL);
@@ -32,7 +46,29 @@ class VertxGraphqlProcessor {
     }
 
     @BuildStep
-    NativeImagePackageResourceBuildItem registerNativeImageResources() {
-        return new NativeImagePackageResourceBuildItem("io/vertx/ext/web/handler/graphiql");
+    @Record(ExecutionTime.STATIC_INIT)
+    void registerVertxGraphqlUI(VertxGraphqlRecorder recorder,
+            BuildProducer<NativeImageResourceDirectoryBuildItem> nativeResourcesProducer, VertxGraphqlConfig config,
+            LaunchModeBuildItem launchMode, BuildProducer<NotFoundPageDisplayableEndpointBuildItem> displayableEndpoints,
+            BuildProducer<RouteBuildItem> routes) {
+
+        boolean includeVertxGraphqlUi = launchMode.getLaunchMode().isDevOrTest() || config.ui.alwaysInclude;
+        if (!includeVertxGraphqlUi) {
+            return;
+        }
+
+        Matcher matcher = TRAILING_SLASH_SUFFIX_REGEX.matcher(config.ui.path);
+        String path = matcher.replaceAll("");
+        if (path.isEmpty()) {
+            throw new ConfigurationException(
+                    "quarkus.vertx-graphql.ui.path was set to \"" + config.ui.path
+                            + "\", this is not allowed as it blocks the application from serving anything else.");
+        }
+
+        Handler<RoutingContext> handler = recorder.handler(path);
+        routes.produce(new RouteBuildItem(path, handler));
+        routes.produce(new RouteBuildItem(path + "/*", handler));
+        displayableEndpoints.produce(new NotFoundPageDisplayableEndpointBuildItem(path + "/"));
+        nativeResourcesProducer.produce(new NativeImageResourceDirectoryBuildItem("io/vertx/ext/web/handler/graphiql"));
     }
 }

--- a/extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlProcessor.java
+++ b/extensions/vertx-graphql/deployment/src/main/java/io/quarkus/vertx/graphql/deployment/VertxGraphqlProcessor.java
@@ -1,11 +1,18 @@
 package io.quarkus.vertx.graphql.deployment;
 
+import java.util.Arrays;
+import java.util.List;
+
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.NativeImagePackageResourceBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.vertx.http.deployment.WebsocketSubProtocolsBuildItem;
+import io.vertx.ext.web.handler.graphql.impl.GraphQLBatch;
+import io.vertx.ext.web.handler.graphql.impl.GraphQLInputDeserializer;
+import io.vertx.ext.web.handler.graphql.impl.GraphQLQuery;
 
 class VertxGraphqlProcessor {
-
     @BuildStep
     FeatureBuildItem feature() {
         return new FeatureBuildItem(FeatureBuildItem.VERTX_GRAPHQL);
@@ -14,5 +21,18 @@ class VertxGraphqlProcessor {
     @BuildStep
     WebsocketSubProtocolsBuildItem websocketSubProtocols() {
         return new WebsocketSubProtocolsBuildItem("graphql-ws");
+    }
+
+    @BuildStep
+    List<ReflectiveClassBuildItem> registerForReflection() {
+        return Arrays.asList(
+                new ReflectiveClassBuildItem(true, true, GraphQLInputDeserializer.class.getName()),
+                new ReflectiveClassBuildItem(true, true, GraphQLBatch.class.getName()),
+                new ReflectiveClassBuildItem(true, true, GraphQLQuery.class.getName()));
+    }
+
+    @BuildStep
+    NativeImagePackageResourceBuildItem registerNativeImageResources() {
+        return new NativeImagePackageResourceBuildItem("io/vertx/ext/web/handler/graphiql");
     }
 }

--- a/extensions/vertx-graphql/deployment/src/test/java/io/quarkus/vertx/graphql/deployment/ErroneousConfigTest.java
+++ b/extensions/vertx-graphql/deployment/src/test/java/io/quarkus/vertx/graphql/deployment/ErroneousConfigTest.java
@@ -1,0 +1,25 @@
+package io.quarkus.vertx.graphql.deployment;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.runtime.configuration.ConfigurationException;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ErroneousConfigTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setExpectedException(ConfigurationException.class)
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addAsResource(new StringAsset("quarkus.vertx-graphql.ui.path=/\n"), "application.properties"));
+
+    @Test
+    public void shouldNotStartApplicationIfPathIsASlash() {
+        Assertions.fail();
+    }
+}

--- a/extensions/vertx-graphql/deployment/src/test/java/io/quarkus/vertx/graphql/deployment/ServingUIFromCustomPathTest.java
+++ b/extensions/vertx-graphql/deployment/src/test/java/io/quarkus/vertx/graphql/deployment/ServingUIFromCustomPathTest.java
@@ -1,0 +1,23 @@
+package io.quarkus.vertx.graphql.deployment;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+public class ServingUIFromCustomPathTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addAsResource(new StringAsset("quarkus.vertx-graphql.ui.path=/custom\n"), "application.properties"));
+
+    @Test
+    public void shouldServeVertxGraphqlUiFromCustomPath() {
+        RestAssured.when().get("/custom").then().statusCode(200);
+    }
+}

--- a/extensions/vertx-graphql/deployment/src/test/java/io/quarkus/vertx/graphql/deployment/ServingUIFromDefaultPathTest.java
+++ b/extensions/vertx-graphql/deployment/src/test/java/io/quarkus/vertx/graphql/deployment/ServingUIFromDefaultPathTest.java
@@ -1,0 +1,21 @@
+package io.quarkus.vertx.graphql.deployment;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+public class ServingUIFromDefaultPathTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class));
+
+    @Test
+    public void shouldServeVertxGraphqlUiFromDefaultPath() {
+        RestAssured.when().get("/graphql-ui").then().statusCode(200);
+    }
+}

--- a/extensions/vertx-graphql/runtime/src/main/java/io/quarkus/vertx/graphql/runtime/VertxGraphqlRecorder.java
+++ b/extensions/vertx-graphql/runtime/src/main/java/io/quarkus/vertx/graphql/runtime/VertxGraphqlRecorder.java
@@ -1,0 +1,34 @@
+package io.quarkus.vertx.graphql.runtime;
+
+import io.quarkus.runtime.annotations.Recorder;
+import io.vertx.core.Handler;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.graphql.GraphiQLHandler;
+import io.vertx.ext.web.handler.graphql.GraphiQLHandlerOptions;
+
+@Recorder
+public class VertxGraphqlRecorder {
+    public Handler<RoutingContext> handler(String path) {
+
+        GraphiQLHandlerOptions options = new GraphiQLHandlerOptions();
+        options.setEnabled(true);
+
+        Handler<RoutingContext> handler = GraphiQLHandler.create(options);
+
+        return new Handler<RoutingContext>() {
+            @Override
+            public void handle(RoutingContext event) {
+                if (event.normalisedPath().length() == path.length()) {
+
+                    event.response().setStatusCode(302);
+                    event.response().headers().set(HttpHeaders.LOCATION, path + "/");
+                    event.response().end();
+                    return;
+                }
+
+                handler.handle(event);
+            }
+        };
+    }
+}

--- a/integration-tests/vertx-graphql/src/test/java/io/quarkus/vertx/graphql/it/VertxGraphqlTest.java
+++ b/integration-tests/vertx-graphql/src/test/java/io/quarkus/vertx/graphql/it/VertxGraphqlTest.java
@@ -6,10 +6,10 @@ import static org.hamcrest.CoreMatchers.is;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.TimeUnit;
 
-import javax.inject.Inject;
-
 import org.eclipse.microprofile.config.ConfigProvider;
+import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 
 import io.quarkus.test.junit.QuarkusTest;
@@ -20,12 +20,20 @@ import io.vertx.core.http.WebSocketConnectOptions;
 
 @QuarkusTest
 class VertxGraphqlTest {
-
-    @Inject
-    Vertx vertx;
-
     public static int getPortFromConfig() {
         return ConfigProvider.getConfig().getOptionalValue("quarkus.http.test-port", Integer.class).orElse(8081);
+    }
+
+    private static Vertx vertx;
+
+    @BeforeAll
+    public static void initializeVertx() {
+        vertx = Vertx.vertx();
+    }
+
+    @AfterAll
+    public static void closeVertx() {
+        vertx.close();
     }
 
     @Test

--- a/integration-tests/vertx-graphql/src/test/java/io/quarkus/vertx/graphql/it/VertxGraphqlTest.java
+++ b/integration-tests/vertx-graphql/src/test/java/io/quarkus/vertx/graphql/it/VertxGraphqlTest.java
@@ -4,6 +4,7 @@ import static io.restassured.RestAssured.given;
 import static org.hamcrest.CoreMatchers.is;
 
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 
 import org.eclipse.microprofile.config.ConfigProvider;
@@ -32,8 +33,10 @@ class VertxGraphqlTest {
     }
 
     @AfterAll
-    public static void closeVertx() {
-        vertx.close();
+    public static void closeVertx() throws InterruptedException {
+        CountDownLatch latch = new CountDownLatch(1);
+        vertx.close((h) -> latch.countDown());
+        latch.await();
     }
 
     @Test

